lib/pmc_miller/puppetdb/queue_depth.rb

```diff
@@ -0,0 +1,122 @@
+# frozen_string_literal: true
+
+module PmcMiller
+  ##
+  # Namespace analyzers specific to PuppetDB
+  module PuppetDB
+    ##
+    # Analyze queue_depth data
+    class QueueDepth
+      attr_reader :data
+
+      ##
+      # Creates a new QueueDepth analyzer
+      #
+      # @param data [Array] queue_depth data to analyze
+      #
+      # @return [PmcMiller::PuppetDB::QueueDepth]
+      #
+      def initialize(data)
+        @data = data
+      end
+
+      ##
+      # Calculate rate of change for data.
+      #
+      # The average rate of change is calculated between the first and second
+      # halves of the data set.
+      #
+      # @return [Float]
+      #
+      def rate_of_change
+        return @_change if defined?(@_change)
+
+        @_change = (first_half_slope + last_half_slope) / 2.0
+      end
+
+      ##
+      # Calculate slope for the first half of data.
+      #
+      # The average rate of change (slope) is calculated for the first half of
+      # the data set.
+      #
+      # @return [Float]
+      #
+      def first_half_slope
+        slope(0, @data.length / 2)
+      end
+
+      ##
+      # Calculate slope for the last half of data.
+      #
+      # The average rate of change (slope) is calculated for the last half of
+      # the data set.
+      #
+      # @return [Float]
+      #
+      def last_half_slope
+        slope(@data.length / 2, -1)
+      end
+
+      ##
+      # Calculate slope given points in the data.
+      #
+      # The average rate of change (slope) is calculated for the given points
+      # in the data set.  The data set must be sorted by time object from low
+      # to high.
+      #
+      # @param dpoint1 [Integer] Array index
+      # @param dpoint2 [Integer] Array index
+      #
+      # @return [Float]
+      #
+      def slope(dpoint1, dpoint2)
+        @data = @data.sort_by(&:time)
```

Comment:
hm, this means the data will get sorted at least twice... (slope is currently called twice).  That seems bad, though maybe if it is already sorted the cost to sort the same way is really low?

Reply:
This has been moved into the initialize method of the analyzer object.